src/PostgREST/ApiRequest/Types.hs

```diff
@@ -92,6 +98,23 @@ type Cast = Text
 type Alias = Text
 type Hint = Text
 
+
+-- | A TypedField is a field with sufficient information to be read from JSON with `json_to_recordset`. In a future PR
+-- type fields will also be key for the data representations feature.
+-- Typed fields should have a nonempty field name and a nonempty type. We enforce this invariant in the `resolveField`
+-- function and keep the real constructor private. (It's not foolproof, you can still construct a `TypedField` using an
+-- existing TypedField and the Haskell `{ tfFieldName = "" }` syntax, so don't do that.)
+data TypedField = PrivateTypedField
+   { tfFieldName :: FieldName
+   , tfJsonPath  :: JsonPath
+   , tfIRType    :: Text
+   } deriving (Eq, Ord)
```

Comment:
Hmm yes. I wanted it near `Field`. We will end up with some analogs such as `Filter` vs `TypedFilter` (which leads to `TypedLogicTree` etc too unfortunately). It is true though that like the defining difference is sort of that in `ApiRequest` we have kind of raw user defined data and later we've distilled that into augmented, better validated data. So based on that extinction, certainly putting it in `Plan/` makes a lot of sense because it will be `Plan` that's responsible for that distillation.

Reply:
Has now been moved.